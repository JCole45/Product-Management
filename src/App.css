.App {
  display: flex;
  justify-content: center;
  align-items: center;
  max-width: 100%;
  padding: 1em 1em;
}

.icon-delete, .icon-edit {
  cursor: pointer;
}

.icon-edit {
  margin-right: 30px;
}

.icon-delete{
  color: red
}

.button{
  margin-top: 20px !important;
  margin-bottom: 5px !important;
  margin-right: 3px !important;
}

.head {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
}
