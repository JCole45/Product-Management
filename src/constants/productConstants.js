
export const FETCH_PRODUCT_REQUEST = 'FETCH_PRODUCT_REQUEST'
export const FETCH_PRODUCT_SUCCESS = 'FETCH_PRODUCT_SUCCESS'
export const FETCH_PRODUCT_FAIL = 'FETCH_PRODUCT_FAIL'
export const FETCH_PRODUCT_RESET = 'FETCH_PRODUCT_RESET'

export const CREATE_PRODUCT = 'CREATE_PRODUCT'

export const UPDATE_PRODUCT = 'UPDATE_PRODUCT'

export const DELETE_PRODUCT = 'DELETE_PRODUCT'