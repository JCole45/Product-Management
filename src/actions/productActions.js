import axios from 'axios'
import { CREATE_PRODUCT, DELETE_PRODUCT, FETCH_PRODUCT_FAIL, FETCH_PRODUCT_REQUEST, FETCH_PRODUCT_SUCCESS, UPDATE_PRODUCT } from "../constants/productConstants"


export const fetchProduct = () => async (dispatch) => {
    try {
        dispatch({ type: FETCH_PRODUCT_REQUEST })

        const { data } = await axios.get(`http://www.mocky.io/v2/5c3e15e63500006e003e9795`)
        const {products} = data
        
        products.forEach((item) => {
            dispatch({
               type: FETCH_PRODUCT_SUCCESS,
               payload: item
            })
        })
        // dispatch({
        //     type: FETCH_PRODUCT_SUCCESS,
        //     payload: products
        // })
    } catch (err) {
        console.log(err)
        dispatch({
            type: FETCH_PRODUCT_FAIL,
            payload: err
        })
    }
}

export const createProduct = (productInfo) => (dispatch, getState) => {
    dispatch({
        type: CREATE_PRODUCT,
        payload: productInfo
    })

    localStorage.setItem('storePharmacyProducts', JSON.stringify(getState().product.products))

}

export const updateProduct = (productInfo) => (dispatch, getState) => {
    dispatch({
        type: UPDATE_PRODUCT,
        payload: productInfo
    })

    localStorage.setItem('storePharmacyProducts', JSON.stringify(getState().product.products))

}

export const deleteProduct = (id) => (dispatch, getState) => {
    dispatch({
        type: DELETE_PRODUCT,
        payload: id
    })

    localStorage.setItem('storePharmacyProducts', JSON.stringify(getState().product.products))

}