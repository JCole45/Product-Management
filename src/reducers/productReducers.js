import { CREATE_PRODUCT, DELETE_PRODUCT, FETCH_PRODUCT_FAIL, FETCH_PRODUCT_REQUEST, FETCH_PRODUCT_SUCCESS, UPDATE_PRODUCT } from "../constants/productConstants";

const checkProducts = (products, newItem) => {
    for(var i =0; i<products.length; i++){
        if(products[i].id === newItem.id){
            console.log()
            return false
        }
    }
    return true
}

export const productReducer = (state = { products: [], loading: false }, action) => {
    switch (action.type) {
        case FETCH_PRODUCT_REQUEST:
            return { products: state.products, loading: true }
        case FETCH_PRODUCT_SUCCESS:
            //return { products: [...state.products,  ...action.payload], loading: false }
            return checkProducts(state.products, action.payload) ? 
                  {products: [...state.products, action.payload], loading: false} 
                  :
                  {products: state.products, loading:false} 
        case FETCH_PRODUCT_FAIL:
            return { products: [], loading: false, error: action.payload }


        case CREATE_PRODUCT:
            return { products: [...state.products, action.payload], loading: false }


        case UPDATE_PRODUCT:
            return {
                products: state.products.map(product => {
                    return action.payload.id === product.id ?
                        {
                            id: action.payload.id,
                            name: action.payload.name, 
                            prices: action.payload.prices.price === product.prices[0].price ? product.prices : [...product.prices, action.payload.prices]
                        }
                        : 
                        product
                })
            }


        case DELETE_PRODUCT:
            return {
                products: state.products.map(product => {
                    return action.payload === product.id ?
                        { ...product, hidden: true } : product
                })
            }

        default:
            return state
    }
}