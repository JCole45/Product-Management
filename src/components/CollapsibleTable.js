import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import { makeStyles } from '@material-ui/core/styles';
import { Button } from "@material-ui/core";
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Collapse from '@material-ui/core/Collapse';
import IconButton from '@material-ui/core/IconButton';
import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';

import TextField from '@material-ui/core/TextField';
import Backdrop from '@material-ui/core/Backdrop';
import Fade from '@material-ui/core/Fade';
import Modal from '@material-ui/core/Modal';
import Grid from '@material-ui/core/Grid';
import { v4 as uuidv4 } from 'uuid';
import DeleteOutlinedIcon from '@material-ui/icons/DeleteOutlined';
import EditIcon from '@material-ui/icons/Edit';

import { createProduct, updateProduct, deleteProduct, restoreProduct } from '../actions/productActions'

const useRowStyles = makeStyles({
    root: {
        '& > *': {
            borderBottom: 'unset',
        },
    },
});

const useStyles = makeStyles((theme) => ({
    table: {
        minWidth: 700,
    },
    dropdownLink: {
        padding: theme.spacing(1),
        flexShrink: 0,
        color: 'black !important',
        fontSize: '0.905rem '
    },
    modal: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
    },
    paper: {
        backgroundColor: theme.palette.background.paper,
        border: '2px solid #000',
        boxShadow: theme.shadows[5],
        padding: theme.spacing(2, 4, 3),
        width: '60%',
    },
    headcolor: {
        color:'white'
    },
    cellshadevisible: {
        color: 'black'
    },
    cellshadehidden: {
        color: 'grey'
    },
}));

const CollapsibleTable = (props) => {
    const { row, visible } = props;
    const [open_history, setOpenHistory] = React.useState(false);
    const classes_ = useRowStyles();

    const dispatch = useDispatch()
    const classes = useStyles();

    const [open, setOpen] = useState(false);
    const [state, setState] = useState('')

    const [id, setId] = useState('')
    const [name, setName] = useState('')
    const [price, setPrice] = useState('')

    const editProduct = (item) => {
        setOpen(true);
        setState('edit')
        setName(item.name)
        setPrice(item._prices[0].price)
        setId(item._id)
    };

    const handleModalClose = () => {
        setOpen(false);
    };

    const handleName = (e) => {
        setName(e.target.value)
    }

    const handlePrice = (e) => {
        setPrice(e.target.value)
    }

    const handleDelete = (id) => {
        dispatch(deleteProduct(id))
    }

    const getDate = () => {
        Date.prototype.toIsoString = function () {
            var tzo = -this.getTimezoneOffset(),
                dif = tzo >= 0 ? '+' : '-',
                pad = function (num) {
                    var norm = Math.floor(Math.abs(num));
                    return (norm < 10 ? '0' : '') + norm;
                };
            return this.getFullYear() +
                '-' + pad(this.getMonth() + 1) +
                '-' + pad(this.getDate()) +
                'T' + pad(this.getHours()) +
                ':' + pad(this.getMinutes()) +
                ':' + pad(this.getSeconds()) +
                dif + pad(tzo / 60) +
                ':' + pad(tzo % 60);
        }

        let date = new Date()

        return date.toIsoString()
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        if (state === 'add') {
            dispatch(createProduct({
                id: uuidv4(),
                name: name,
                prices: [{
                    id: uuidv4(),
                    price: Number(price),
                    date: getDate()
                }]
            }))
            setOpen(false);
            setName('')
            setPrice()
        }

        if (state === 'edit') {
            dispatch(updateProduct({
                id: id,
                name: name,
                prices: {
                    id: uuidv4(),
                    price: Number(price),
                    date: getDate()
                }
            }))
            setOpen(false);
            setName('')
            setPrice()
        }
    }

    const sortbyDate = (array) => {
        return array.sort(function (a, b) {
            return new Date(b.date) - new Date(a.date)
        })
    }

    const handleRestore = (id) => {
       dispatch(restoreProduct(id))
    }

    const rowshade = visible ? classes.cellshadevisible : classes.cellshadehidden

    return (
        <React.Fragment>
            <Modal
                aria-labelledby="transition-modal-title"
                aria-describedby="transition-modal-description"
                className={classes.modal}
                open={open}
                onClose={handleModalClose}
                closeAfterTransition
                BackdropComponent={Backdrop}
                BackdropProps={{
                    timeout: 500,
                }}
            >
                <Fade in={open}>
                    <div className={classes.paper}>
                        <h2 id="transition-modal-title">{state === 'add' ? 'Add Item' : state === 'edit' ? 'Edit Item' : ''}</h2>
                        <form onSubmit={handleSubmit}>
                            <Grid item xs={12} sm={5}>
                                <TextField
                                    id="name"
                                    name="name"
                                    label={"Drug name"}
                                    type={"text"}
                                    fullWidth
                                    onChange={handleName}
                                    value={name}
                                />

                                <TextField
                                    id="price"
                                    name="price"
                                    label={"Price"}
                                    type={"number"}
                                    onChange={handlePrice}
                                    value={price}
                                    fullWidth
                                />
                            </Grid>
                            {state === 'add' ?
                                <Button className="button" type="submit" color="primary" variant="outlined" size="small">
                                    Add
                                </Button>
                                :

                                state === 'edit' ?
                                    <Button className="button" type="submit" color="primary" variant="outlined" size="small">
                                        Edit
                                    </Button>
                                    :
                                    ''
                            }
                        </form>
                    </div>
                </Fade>
            </Modal>

            <TableRow className={classes_.root}>
                <TableCell>
                    <IconButton aria-label="expand row" size="small" onClick={() => setOpenHistory(!open_history)}>
                        {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
                    </IconButton>
                </TableCell>
                <TableCell className={rowshade} component="th" scope="row">
                    {row.name}
                </TableCell>
                {/* <TableCell align="right">{row.name}</TableCell> */}
                <TableCell className={rowshade} align="right">GH&cent; {sortbyDate(row._prices)[0].price.toFixed(2)}</TableCell>
                <TableCell className={rowshade} align="right">{new Date(sortbyDate(row._prices)[0].date).toString()}</TableCell>
                <TableCell className={rowshade} align="right">
                    {visible ?
                    <>
                    <EditIcon color="primary" rowshade="icon-edit" onClick={() => editProduct(row)} />
                    <DeleteOutlinedIcon className="icon-delete" onClick={() => handleDelete(row._id)} />
                    </>
                    :
                    <Button onClick={()=> handleRestore(row._id)}  color="primary" variant="contained" size="small">
                        Restore
                    </Button>
                    }
                </TableCell>
            </TableRow>
            <TableRow>
                <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
                    <Collapse in={open_history} timeout="auto" unmountOnExit>
                        <Box margin={1}>
                            <Typography variant="h6" gutterBottom component="div">
                                Price History
                            </Typography>
                            <Table size="small" aria-label="purchases">
                                <TableHead>
                                    <TableRow>
                                        <TableCell>Price</TableCell>
                                        <TableCell>Date</TableCell>
                                    </TableRow>
                                </TableHead>
                                <TableBody>
                                    {row.history.map((historyRow) => (
                                        <TableRow key={historyRow.date}>
                                            <TableCell>GH&cent;{historyRow.price}</TableCell>
                                            <TableCell component="th" scope="row">
                                                {new Date(historyRow.date).toString()}
                                            </TableCell>
                                        </TableRow>
                                    ))}
                                </TableBody>
                            </Table>
                        </Box>
                    </Collapse>
                </TableCell>
            </TableRow>
        </React.Fragment>
    );
}

export default CollapsibleTable
