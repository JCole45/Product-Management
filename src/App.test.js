import { render, screen } from '@testing-library/react';
import App from './App';
import { createProduct, updateProduct } from '../../mpharma-test/src/actions/productActions';
import { CREATE_PRODUCT, UPDATE_PRODUCT, DELETE_PRODUCT } from './constants/productConstants'
import { v4 as uuidv4 } from 'uuid';
import { Provider } from 'react-redux'
import store from './store'
import { mount, shallow } from 'enzyme'
import CollapsibleTable from './components/CollapsibleTable'
import renderer from 'react-test-renderer';
import { productReducer } from './reducers/productReducers'


const getDate = () => {
  Date.prototype.toIsoString = function () {
    var tzo = -this.getTimezoneOffset(),
      dif = tzo >= 0 ? '+' : '-',
      pad = function (num) {
        var norm = Math.floor(Math.abs(num));
        return (norm < 10 ? '0' : '') + norm;
      };
    return this.getFullYear() +
      '-' + pad(this.getMonth() + 1) +
      '-' + pad(this.getDate()) +
      'T' + pad(this.getHours()) +
      ':' + pad(this.getMinutes()) +
      ':' + pad(this.getSeconds()) +
      dif + pad(tzo / 60) +
      ':' + pad(tzo % 60);
  }
}

describe('Table row', () => {
  test('snapshot renders', () => {
    const row = {
      name: 'Test',
      _prices: [{ id: 'id', price: 10, date: getDate() }],
      history: [{ id: 'id', price: 10, date: getDate() }],
      _id: '1',
      visible: true
    }
    const component = renderer.create(<Provider store={store}><CollapsibleTable key={'1'} visible={true} row={row} /></Provider>)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})


describe('reducer', () => {
  it('should return  the initial state', () => {
    expect(productReducer(undefined, { products: [], loading: false })).toEqual({ products: [], loading: false })
  });

  it('should handle create product', () => {
    const itemInfo = {
      id: uuidv4(),
      name: 'Test Product',
      prices: [{
        id: uuidv4(),
        price: 40,
        date: getDate()
      }]
    }

    const action = {
      type: CREATE_PRODUCT,
      payload: itemInfo
    }
    expect(productReducer({ products: [] }, action)).toEqual({loading:false, products:[itemInfo]})
  });

  it('should update product', () => {
    const id = '1_90'
    const itemInfo = {
      id: id,
      name: 'Test Product',
      prices: [{
        id: uuidv4(),
        prices: [{
          id: uuidv4(),
          price: 40,
          date: getDate()
        }],
        date: getDate()
      }]
    }

    const itemInfoUpdate = {
      id: id,
      name: 'Test Product Update',
      prices: [{
        id: uuidv4(),
        prices: [{
          id: uuidv4(),
          price: 40,
          date: getDate()
        }],
        date: getDate()
      }]
    }

    const action = {
      type: UPDATE_PRODUCT,
      payload: itemInfo
    }
    expect(productReducer({ products: [itemInfo], loading:false }, action)).toEqual({loading:false, products:[itemInfoUpdate]})
  })
})

