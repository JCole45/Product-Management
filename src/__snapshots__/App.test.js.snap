// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table row snapshot renders 1`] = `
Array [
  <tr
    className="MuiTableRow-root makeStyles-root-1"
    role={null}
  >
    <td
      aria-sort={null}
      className="MuiTableCell-root"
    >
      <button
        aria-label="expand row"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
        </span>
      </button>
    </td>
    <th
      aria-sort={null}
      className="MuiTableCell-root makeStyles-cellshadevisible-7"
      role="cell"
      scope="row"
    >
      Test
    </th>
    <td
      aria-sort={null}
      className="MuiTableCell-root makeStyles-cellshadevisible-7 MuiTableCell-alignRight"
    >
      GHÂ¢ 
      10.00
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root makeStyles-cellshadevisible-7 MuiTableCell-alignRight"
    >
      Invalid Date
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root makeStyles-cellshadevisible-7 MuiTableCell-alignRight"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
        focusable="false"
        onClick={[Function]}
        rowshade="icon-edit"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root icon-delete"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M16 9v10H8V9h8m-1.5-6h-5l-1 1H5v2h14V4h-3.5l-1-1zM18 7H6v12c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7z"
        />
      </svg>
    </td>
  </tr>,
  <tr
    className="MuiTableRow-root"
    role={null}
  >
    <td
      aria-sort={null}
      className="MuiTableCell-root"
      colSpan={6}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    />
  </tr>,
]
`;
